src/Vector.cpp

```diff
@@ -0,0 +1,349 @@
+*/
+void Vector::insert(size_t index, int value) {
+
+    if (index < 0 || index > m_size)
+        throw std::out_of_range("Index out of range");
+
+
+    //resizing the capacity of the vector
+    else if (m_size == m_capacity) {
+        if (m_capacity == 0)
+            m_capacity = 1;
+
+        m_capacity = m_capacity < 128 ? 2 * m_capacity : m_capacity * 3 / 2;
+
+        int* newData = new int[m_capacity];
+        for (size_t i = 0; i < index; ++i) {
+            newData[i] = m_data[i];
+        }
+        newData[index] = value;
+        for (size_t i = index; i < m_size; ++i) {
+            newData[i + 1] = m_data[i];
+        }
+        delete[] m_data;
+        m_data = newData;
+    }
+
+    else {
+        //moving the values one place right until getting to index received
+        for (size_t i = m_size; i > index; --i) {
+            m_data[i] = m_data[i - 1];
+        }
+        m_data[index] = value;
+    }
+    ++m_size;
+}
+/**
+
+@brief Removes an element at a specific index in the vector.
+@param index The index of the element to remove.
+@throws std::out_of_range if the index is out of range (i.e., less than 0 or greater than or equal to the current size of the vector), or if the vector is empty.
+*/
+void Vector::erase(const size_t index) {
+
+    if (empty())
+        throw std::out_of_range("Vector is empty");
+
+    if (index < 0 || index >= m_size)
+        throw std::out_of_range("Index out of range");
+
+    for (size_t i = index; i < m_size-1; i++)
+        m_data[i] = m_data[i+1];
+
+    m_size--;
+
+}
+/**
+
+@brief Returns a pointer to the internal array of the vector.
+@return int* A pointer to the internal array of the vector.
+@return nullptr if the vector is empty.
+*/
+int *Vector::data() {
+    if (empty())
+        return nullptr;
+
+    return &m_data[0];
+}
+
+/**
+
+@brief Returns whether the vector is empty or not.
+@return true if the vector is empty, false otherwise.
+*/
+bool Vector::empty() const {
+    if (m_size == 0)
+        return true;
+
+    return false;
+}
+
+/**
+
+@brief Returns the size of the vector.
+@return size_t The size of the vector.
+*/
+size_t Vector::size() const {
+    return m_size;
+}
+
+/**
+
+@brief Returns the capacity of the vector.
+@return size_t The capacity of the vector.
+*/
+size_t Vector::capacity() const {
+    return m_capacity;
+}
+/**
+
+@brief Clears the vector and deallocates all memory associated with it.
+*/
+void Vector::clear() {
+    delete[] m_data;
+    m_size = 0;
+    m_capacity = 0;
+    m_data = new int [m_capacity];
+}
+/**
+
+@brief Resizes the vector to a new capacity and fills new elements with a given value.
+@param newCapacity The new capacity of the vector.
+@param value The value to be assigned to new elements.
+*/
+void Vector::resize(const size_t newCapacity, const int value) {
+    std::cout << "value is: " << value;
+    int* newData = new int[newCapacity];
```

Comment:
Are you sure it is required to allocate a new array? What if the requested size is smaller than the current capacity?
You could spare an allocation (which is considered a heavy operation) this way